### PR TITLE
Feature/2

### PR DESCRIPTION
**Pull request by julySteuer**

### CODE REVIEW

File: src/songuploader/utils.py [hunk 1]
```diff
@@ -17,25 +17,30 @@
 from uploader.models import Submission
 import os 
 
-from songuploader.settings import MEDIA_ROOT
-from uploader.models import Submission
-
-
 class ConfiguredLoginViewMixin(LoginRequiredMixin):
     login_url = reverse_lazy("login")
     redirect_field_name = "redirect_to"
```
Comment (SFSeeger) [inline src/songuploader/utils.py:23]: can create errors with unusual usernames containing dots or spaces

File: src/songuploader/utils.py [hunk 2]
```diff
@@ -17,25 +17,30 @@
...
+
 def download_song(submission: Submission):
-    out_path = os.path.join("/tmp", f"{submission.user.username}.mp4")
-    YouTube(url=submission.song_url).streams.filter(only_audio=True).first().download(output_path=os.path.dirname(out_path), filename=os.path.basename(out_path))
-    open_file = open(out_path, 'rb')
+    mp4_out_path = os.path.join("/tmp", f"{submission.user.username}.mp4")
+    mp3_out_path = os.path.join("/tmp", f"{submission.user.username}.mp3")
+    YouTube(url=submission.song_url).streams.filter(only_audio=True).first().download(output_path=os.path.dirname(mp4_out_path), filename=os.path.basename(mp4_out_path))
+    os.system(f"ffmpeg -i {mp4_out_path} -vn {mp3_out_path}")
+    open_file = open(mp3_out_path, 'rb')
```
Comment (SFSeeger) [inline src/songuploader/utils.py:33]: please use the native copy of os

File: src/songuploader/utils.py (hunk 1 above)
Comment (SFSeeger) [inline src/songuploader/utils.py:23]: Can lead to errors on unusual filenames containing dots or spaces

File: src/songuploader/utils.py [hunk 3]
```diff
@@ -17,25 +17,30 @@
...
     song_file: File = File(open_file)
     submission.song = song_file
     submission.save()
```
Comment (SFSeeger) [inline src/songuploader/utils.py:36]: Submission needs to be updated, right?

**Review by SFSeeger — CHANGES_REQUESTED** [review]
Unnecessary requirements in src. Otherwise see comments

Please also use the update() method to modify multiple fields on model

This review was made on Github Mobile and could be absolute trash

File: src/songuploader/utils.py [hunk 4]
```diff
@@ -17,25 +17,30 @@
...
-    
+
 def download_song(submission: Submission):
-    out_path = os.path.join("/tmp", f"{submission.user.username}.mp4")
-    YouTube(url=submission.song_url).streams.filter(only_audio=True).first().download(output_path=os.path.dirname(out_path), filename=os.path.basename(out_path))
-    open_file = open(out_path, 'rb')
+    mp4_out_path = os.path.join("/tmp", f"{submission.user.username}.mp4")
+    mp3_out_path = os.path.join("/tmp", f"{submission.user.username}.mp3")
+    YouTube(url=submission.song_url).streams.filter(only_audio=True).first().download(output_path=os.path.dirname(mp4_out_path), filename=os.path.basename(mp4_out_path))
+    os.system(f"ffmpeg -i {mp4_out_path} -vn {mp3_out_path}")
```
Comment (SFSeeger) [inline src/songuploader/utils.py:65]: This can create errors on unusual usernames e.g. containing dots or spaces

File: src/songuploader/utils.py [hunk 5]
```diff
@@ -17,25 +17,30 @@
...
+    open_file = open(mp3_out_path, 'rb')
     song_file: File = File(open_file)
     submission.song = song_file
     submission.save()
+    os.remove(mp4_out_path)
+    os.remove(mp3_out_path)
+
+def slice_song(submission: Submission):
+    out_path = os.path.join("/tmp", os.path.basename(submission.song.name))
+    os.system(f"cp {submission.song.path} {out_path}")
```
Comment (SFSeeger) [inline src/songuploader/utils.py]: Use native `os.copy()`

File: src/songuploader/utils.py [hunk 6]
```diff
@@ -17,25 +17,30 @@
...
+    os.system(f"ffmpeg -ss {submission.start_time} -i {out_path} -c copy -y -t 30 {submission.song.path}")
     os.remove(out_path)
```
Comment (SFSeeger) [inline src/songuploader/utils.py:81]: Test if submission has to be updated

File: src/songuploader/utils.py [hunk 7]
```diff
@@ -17,25 +17,30 @@
...
     song_file: File = File(open_file)
     submission.song = song_file
     submission.save()
+    os.remove(mp4_out_path)
+    os.remove(mp3_out_path)
+
+def slice_song(submission: Submission):
+    out_path = os.path.join("/tmp", os.path.basename(submission.song.name))
+    os.system(f"cp {submission.song.path} {out_path}")
+    os.system(f"ffmpeg -ss {submission.start_time} -i {out_path} -c copy -y -t 30 {submission.song.path}")
```
Comment (SFSeeger) [inline src/songuploader/utils.py]: Use `end_time` variable to determine length

**Review by SFSeeger — CHANGES_REQUESTED** [review]

File: src/songuploader/utils.py (hunk 7 above)
Comment (julySteuer) [reply src/songuploader/utils.py]: Oh i thought we alwys do 30 seconds

File: src/songuploader/utils.py (hunk 6 above)
Comment (julySteuer) [reply src/songuploader/utils.py:81]: What exactly. The file name and location saty the same so there is no need to update the submission. There can be some django magic behind it which breaks this stuff but it works right now i can download the song. slice it up using the submission and play the sliced version of it

File: src/songuploader/utils.py (hunk 4 above)
Comment (julySteuer) [reply src/songuploader/utils.py:65]: Should i use the userid instead. Are there usernames where such behaviour is present

File: src/songuploader/utils.py (hunk 5 above)
Comment (julySteuer) [reply src/songuploader/utils.py]: I am really smooth brain lol

Comment (julySteuer) [reply src/songuploader/utils.py]: Oh wait right there is no os.copy

Comment (SFSeeger) [reply src/songuploader/utils.py]: True it is part of the build in [shutil](https://docs.python.org/3/library/shutil.html#module-shutil) package. It also has multiple copy types depending on usecase.
See [here](https://stackoverflow.com/questions/123198/how-to-copy-files)

File: src/songuploader/utils.py (hunk 6 above)
Comment (SFSeeger) [reply src/songuploader/utils.py:81]: Ok, perfect

File: src/songuploader/utils.py (hunk 7 above)
Comment (SFSeeger) [reply src/songuploader/utils.py]: It's the goal, but i would reccoment to use the endtime

File: src/songuploader/utils.py (hunk 4 above)
Comment (SFSeeger) [reply src/songuploader/utils.py:65]: 👍

File: src/songuploader/utils.py (hunk 7 above)
Comment (julySteuer) [reply src/songuploader/utils.py]: If the end time is set by the user there is the posibility for an exploid

File: src/songuploader/utils.py (hunk 5 above)
Comment (julySteuer) [reply src/songuploader/utils.py]: Now using shutil

**Review by SFSeeger — APPROVED** [review]

**Review by julySteuer — COMMENTED** [review]
👍

**Pull request merged**